### PR TITLE
Handle un-pickleable exceptions

### LINKED ISSUE #18: Fix directories so that the conda command-line utils have right path
- [ ] Are the conda utils / is the anaconda environment in the runtime? 
- [ ] Can we conda install or pip install something?

### PR DESCRIPTION
**Pull request by ericmjonas**
While attempting to resolve #18 I discovered that subprocess would raise an exception when invoked with a non-existant command, but that things would then fail when unpickling. Closer investigation showed that some custom python exceptions don't invoke the base class constructor and as a result they can't be unpickled. Now, pickling exceptions is a weird thing, so I guess I understand why this happens. 

To resolve, when an exception is triggered, `jobrunner.py` tries to save it and then reload it. If this fails we then save string versions of the exception and do some cleanup on the client side. It's not ideal, but an exception still propagates, which is what matters. 

Note this PR is a branch from  `release-setup` and should only be merged after.

### CODE REVIEW

File: pywren/jobrunner.py [hunk 1]
```diff
@@ -38,11 +38,34 @@ def b64str_to_bytes(str_data):
...
+                    open(out_filename, 'wb'), -1)
+        # this is just to make sure they can be unpickled
+        pickle.load(open(out_filename, 'rb'))
+
+    except Exception as pickle_exception:
+        pickle.dump({'result' : str(e), 
+                     'exc_type' : str(exc_type), 
+                     'exc_value' : str(exc_value), 
+                     'exc_traceback' : exc_traceback, 
+                     'exc_traceback_str' : str(exc_traceback),
```
Comment (shivaram) [inline pywren/jobrunner.py]: do we need `exc_traceback` and `exec_traceback_str` ? also does `exc_traceback` always get pickled correctly ?

File: pywren/wren.py [hunk 2]
```diff
@@ -539,6 +539,7 @@ def result(self, timeout=None, check_only=False, throw_except=True):
                 raise Exception(exception_str, *exception_args)
         
         call_output_time = time.time()
+        print("calling get_call_output")
```
Comment (shivaram) [inline pywren/wren.py]: is this useful to print each time or just something useful for debugging ?

File: pywren/wren.py [hunk 3]
```diff
@@ -556,28 +557,36 @@ def result(self, timeout=None, check_only=False, throw_except=True):
...
+
+
         if call_success:
 
             self._return_val = call_invoker_result['result']
             self._state = JobState.success
-        else:
+            return self._return_val
+
+        elif call_success == False and throw_except:
```
Comment (shivaram) [inline pywren/wren.py]: `call_success` is always False here - so this can be `if throw_except` (since you return from previous if)

File: pywren/wren.py [hunk 4]
```diff
@@ -556,28 +557,36 @@ def result(self, timeout=None, check_only=False, throw_except=True):
...
+
             self._exception = call_invoker_result['result']
             self._traceback = (call_invoker_result['exc_type'], 
                                call_invoker_result['exc_value'], 
                                call_invoker_result['exc_traceback'])
 
             self._state = JobState.error
+            if call_invoker_result.get('pickle_fail', False):
+                logging.warning("there was an error pickling the resulting exception: {}".format(call_invoker_result['exc_value']))
```
Comment (shivaram) [inline pywren/wren.py]: minor: but it might also be worth while printing the pickle_exception here

File: pywren/jobrunner.py [hunk 5]
```diff
@@ -38,11 +38,34 @@ def b64str_to_bytes(str_data):
...
+    try:
+        pickle.dump({'result' : e, 
+                     'exc_type' : exc_type, 
+                     'exc_value' : exc_value, 
+                     'exc_traceback' : exc_traceback, 
+                     'sys.path' : sys.path, 
+                     'success' : False}, 
+                    open(out_filename, 'wb'), -1)
+        # this is just to make sure they can be unpickled
+        pickle.load(open(out_filename, 'rb'))
```
Comment (ooq) [inline pywren/jobrunner.py]: should make sure fp for `open(out_filename, 'wb')` is closed before the load?

File: pywren/jobrunner.py (hunk 1 above)
Comment (ericmjonas) [inline pywren/jobrunner.py]: `exc_traceback` seems to be pickled on the server side much more reliably, but may still fail on the unpickling -- thus wanting a string copy of it that we can use no matter what.

File: pywren/wren.py (hunk 2 above)
Comment (ericmjonas) [inline pywren/wren.py]: thanks, debugging detritus

File: pywren/jobrunner.py (hunk 1 above)
Comment (shivaram) [inline pywren/jobrunner.py]: Then it'd be safer to not include the traceback and just keep the traceback str ? I just worry we will have cascading failures

**Comment by ooq** [comment]
Is this issue resolved? @ericmjonas 
I'm asking because I'm running into this problem. I guess to make things simple, we can just stringfy all of these objects that have potential pickle problems? The only benefit of keeping them as objects seem to be that we can reraise the exception later?

**Comment by ericmjonas** [comment]
Ha, it was supposed to be. I think this is always going to be a challenge. I had thought we had chased out all the edge cases here, but it's hard to test. 

I really like the user experience of being able to reraise the exception and give a sane traceback, and have found it quite useful. For stability this might be something we ultimately have to abandon, but I hope not. 

Is there any chance you have some captured output /  logs from when this happens, or better yet a reproducible test case? :)

**Comment by ericmjonas** [comment]
Also Github won't let me reopen the issue because it was on a branch? I'm going to create a new issue for us to keep tracking this.
